### PR TITLE
1077 ensure reproducible app ids

### LINKED ISSUE #1077: [Bug]: snapshot manager json `app_id` does match with app.
### What happened?

 when the app is disconnected from the server, the `app_id `in the downloaded snapshot file's `JSON ` differerent from the app's `app_id`. I used the JSON locally and  encountered same issue.it gives message, "`This snapshot file is not compatible with the app and cannot be loaded."`

Here demo:

https://github.com/insightsengineering/teal/assets/6700955/71a9e9ed-af12-458d-8b7b-e54ef143fe87

### PR DESCRIPTION
**Pull request by chlebowa**
Closes #1077 

Added a utility function for use in `create_app_id`:
- `defunction` inspects list and converts functions to strings representing their bodies, recursively

Modified `create_app_id` to use deparsed function body of `teal_data_module$server` rather than the language object.

#### TESTING
Please run this code on your machine, using this branch (including running and killing an app). 
Report the result of `create_app_id`. Hopefully, both calls will produce the same hash.
```
devtools::load_all("../teal")
library(teal.modules.general)

td <- teal_data() %>%
  within(library("magrittr")) %>% within(library(cond)) %>% within(iris <- iris) %>% within(mtcars <- mtcars)
datanames(td) <- c("iris", "mtcars")

mods <- modules(
  tm_g_distribution(
    dist_var = teal.transform::data_extract_spec(
      dataname = "iris",
      select = teal.transform::select_spec(variable_choices("iris"), "Petal.Length"))
  ),
  tm_data_table(
    "table",
    variables_selected = list(),
    dt_args = list()
  ),
  modules(
    label = "tab1",
    teal::example_module("example1", datanames = NULL),
    teal::example_module("example2", datanames = "iris")
  )
)

app <- init(
  data = td,
  modules = mods
)

create_app_id(td, mods)
runApp(app, launch.browser = TRUE)

create_app_id(td, mods)
```

### CODE REVIEW

**Comment by chlebowa** [comment]
Hash is stable.
```
> create_app_id(td, mods)
[1] "03b71cde3884a6cc34c9cf925708f107"
```

**Comment by m7pr** [comment]
Hey, on Windows I had 

```r
"8642db0d3563cdc2614b6f1c91fb83e7"
```

**Comment by vedhav** [comment]
Interesting @m7pr. I do get `03b71cde3884a6cc34c9cf925708f107` as expected on my mac.
Which R and rlang version do you use?
Mine is `rlang v1.1.3` and `R 4.3.2`

**Comment by averissimo** [comment]
I have the same hash as @m7pr 

`rlang 1.1.3` and `R 4.3.2` (Linux)

<details>

<summary>Session info</summary>

```r
R version 4.3.2 (2023-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 23.10

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.11.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.11.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_IE.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_IE.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_IE.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_IE.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/Berlin
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] teal_0.14.0.9042                 cond_1.2-3.1                     survival_3.5-7                  
 [4] statmod_1.5.0                    magrittr_2.0.3                   teal.modules.general_0.2.16.9014
 [7] shinyTree_0.3.1                  ggmosaic_0.3.3                   ggplot2_3.4.4                   
[10] testthat_3.2.1                   teal.slice_0.4.0.9040            teal.data_0.4.0.9000            
[13] teal.code_0.5.0.9002             shiny_1.8.0                     

loaded via a namespace (and not attached):
  [1] polynom_1.4-1             remotes_2.4.2.1           logger_0.2.2             
  [4] rlang_1.1.3               compiler_4.3.2            systemfonts_1.0.5        
  [7] callr_3.7.3               vctrs_0.6.4               quantreg_5.97            
 [10] stringr_1.5.1             profvis_0.3.8             pkgconfig_2.0.3          
 [13] crayon_1.5.2              fastmap_1.1.1             backports_1.4.1          
 [16] ellipsis_0.3.2            labeling_0.4.3            fontawesome_0.5.2        
 [19] utf8_1.2.4                promises_1.2.1            rmarkdown_2.25           
 [22] sessioninfo_1.2.2         ps_1.7.5                  ragg_1.2.6               
 [25] MatrixModels_0.5-3        purrr_1.0.2               xfun_0.41                
 [28] shinyvalidate_0.1.3       cachem_1.0.8              teal.reporter_0.2.1.9015 
 [31] jsonlite_1.8.8            goftest_1.2-3             later_1.3.1              
 [34] broom_1.0.5               prettyunits_1.2.0         R6_2.5.1                 
 [37] bslib_0.6.1               stringi_1.8.1             pkgload_1.3.3            
 [40] jquerylib_0.1.4           brio_1.1.3                Rcpp_1.0.11              
 [43] knitr_1.45                usethis_2.2.2             teal.logger_0.1.3.9010   
 [46] httpuv_1.6.13             Matrix_1.6-3              splines_4.3.2            
 [49] tidyselect_1.2.0          rstudioapi_0.15.0         yaml_2.3.8               
 [52] miniUI_0.1.1.1            processx_3.8.2            teal.widgets_0.4.2.9002  
 [55] pkgbuild_1.4.2            lattice_0.22-5            tibble_3.2.1             
 [58] withr_3.0.0               evaluate_0.23             desc_1.4.2               
 [61] ggpp_0.5.6                urlchecker_1.0.1          shinycssloaders_1.0.0    
 [64] pillar_1.9.0              BiocManager_1.30.22       teal.transform_0.4.0.9011
 [67] DT_0.30                   checkmate_2.3.1           renv_1.0.3               
 [70] shinyjs_2.1.0             plotly_4.10.3             generics_0.1.3           
 [73] rprojroot_2.0.4           munsell_0.5.0             scales_1.2.1             
 [76] xtable_1.8-4              glue_1.6.2                lazyeval_0.2.2           
 [79] tools_4.3.2               ggpmisc_0.5.5             data.table_1.14.8        
 [82] SparseM_1.81              fs_1.6.3                  grid_4.3.2               
 [85] tidyr_1.3.0               crosstalk_1.2.0           devtools_2.4.5           
 [88] colorspace_2.1-0          cli_3.6.1                 textshaping_0.3.7        
 [91] fansi_1.0.5               viridisLite_0.4.2         dplyr_1.1.4              
 [94] gtable_0.3.4              sass_0.4.7                digest_0.6.33            
 [97] ggrepel_0.9.4             farver_2.1.1              htmlwidgets_1.6.2        
[100] memoise_2.0.1             htmltools_0.5.7           lifecycle_1.0.4          
[103] shinyWidgets_0.8.0        httr_1.4.7                mime_0.12                
[106] MASS_7.3-60              
```

</details>

**Comment by averissimo** [comment]
Are you guys running the examples as is? for me I need to prepend `teal.transform::` to line 13.

Although I get the same hash with `library(teal.transform)` on top of the script.

**Comment by averissimo** [comment]
Nevertheless, the hash is consistent within an installation :shrug:

**Comment by vedhav** [comment]
`teal.modules.general_0.2.16.9014` is the black sheep.
On updating it to `.9018` you should get the new hash.

**Comment by chlebowa** [comment]
Confirmed, aligning the versions of packages that encode modules produces the same hash.

**Comment by chlebowa** [comment]
It looks like we got an extra check on session integrity for free :joy:

**Comment by chlebowa** [comment]
Seriously, though, this is a problem. A snapshot should remain valid when an update to, say, `tmc`, is released that changes the internal code of a module.

**Comment by vedhav** [comment]
Ah! True, In that case, our fundamental logic should only include hashing the data and not even include the modules in the hash.

**Comment by averissimo** [comment]
Depends on whether the module changes internally.  If there are changes to the logic / UI, then the App ID should not be the same.

**Comment by vedhav** [comment]
I would argue even if the module changes without any change to the input data, The users should be able to re-use their shanshots. Because they are most likely interested in the same data snapshots. So, I feel that updates to the modules should not affect the snapshots.

**Comment by averissimo** [comment]
That being said, changes in the modules should not change, but maybe the `filter = `, should be included when hashing the data for the snapshot manager (instead of the modules)

**Comment by chlebowa** [comment]
Slow down.

**Comment by vedhav** [comment]
> but maybe the filter = , should be included when hashing

I see your point, If the global and local filter stats are changed it could cause errors in applying the snapshots. I think only this part needs to be hashed.

**Comment by chlebowa** [comment]
1. Snapshot refers to filter state, not data state.
2. We stamp snapshots with app id to ensure that the snapshot is compatible with the app, i.e. the same filters can be created and mapped appropriately.
3. The app id must depend on the data present and modules available.

Now, internal changes to modules will be ignored (meaning, if module `tm_c_mod1` uses an internal function `inter` and that function's body changes, the hash will not be affected. However, any syntactic change will affect the hash, even a logically inert one, like replacing `<-` with `=`.

**Comment by chlebowa** [comment]
Merging after oral approval from the team.

**Comment by chlebowa** [comment]
Can somebody actually review the code? :wink: ?

File: R/utils.R [hunk 1]
```diff
@@ -299,7 +299,34 @@ create_app_id <- function(data, modules) {
...
+
   rlang::hash(list(data = data, modules = modules))
 }
+
+#' Flatten `teal_modules` to a list of `teal_module`s.
+#' @keywords internal
+#' @noRd
+flatten_modules <- function(modules) {
+  if (inherits(modules, "teal_module")) {
+    list(modules)
```
Comment (averissimo) [inline R/utils.R:314]: This will never be reached due to `checkmate::assert_class(modules, "teal_modules")` above

File: R/utils.R [hunk 2]
```diff
@@ -299,7 +299,34 @@ create_app_id <- function(data, modules) {
...
+  } else if (inherits(modules, "teal_modules")) {
+    unlist(lapply(modules$children, flatten_modules), recursive = FALSE)
+  } else {
+    stop("flatten_modules accepts only 'teal_module' and 'teal_modules' objects")
+  }
+}
```
Comment (averissimo) [inline R/utils.R:320]: Do you need to flatten them? They are both lists.

Comment (averissimo) [reply R/utils.R:320]: From my POV it seems it just discards the root label.

File: R/utils.R (hunk 1 above)
Comment (chlebowa) [reply R/utils.R:314]: `teal_modules` can have `teal_modules` elements as well as `teal_module`.

File: R/utils.R [hunk 3]
```diff
@@ -299,7 +299,34 @@ create_app_id <- function(data, modules) {
   data <- if (inherits(data, "teal_data")) {
     as.list(data@env)
   } else if (inherits(data, "teal_data_module")) {
-    body(data$server)
+    deparse1(body(data$server))
   }
+  modules <- lapply(flatten_modules(modules), function(x) lapply(x, defunction))
```
Comment (averissimo) [inline R/utils.R:304]: Given that modules always has `teal_modules` class due to assertion above, this is safe to do

```suggestion
  modules <- lapply(
    unlist(lapply(modules$children, flatten_modules), recursive = FALSE),
    defunction
  )
```

Or

```suggestion
  modules <- lapply(modules, defunction)
```

File: R/utils.R (hunk 1 above)
Comment (averissimo) [reply R/utils.R:314]: Ah you're right. I didn't see the recursion here.

Why visit all elements twice though?

Comment (averissimo) [reply R/utils.R:314]: Unless the idea is to make the App ID resilient to name changes, it seems the same can be achieved by visiting all the elements just once.

The `defunction` will visit all `teal_module` and `teal_modules`

```r
> is.list(modules(example_module()))
[1] TRUE
> is.list(example_module())
[1] TRUE
> 
```

```r
mods <- modules(
  label = "tab1",
  modules(
    modules(
      teal::example_module("example1", datanames = NULL),
      modules(
        modules(
          teal::example_module("example1", datanames = NULL)
        )
      )
    )
  ),
  teal::example_module("example2", datanames = "iris")
)

defunction(mods)
```

Comment (chlebowa) [reply R/utils.R:314]: You're right, `defunction` already handles a nested list.
Removing.

File: tests/testthat/test-utils.R [hunk 4]
```diff
@@ -100,3 +100,104 @@ test_that("build_app_title builts a valid tag", {
...
+  testthat::expect_failure(testthat::expect_identical(hash1, hash2))
+})
+
+## defunction ----
+testthat::test_that("defunction returns a string when passed a function", {
+  testthat::expect_true(
+    checkmate::test_string(
+      defunction(init)
+    )
+  )
```
Comment (averissimo) [inline tests/testthat/test-utils.R:156]:
```suggestion
  checkmate::expect_string(
    defunction(init)
  )
```

File: tests/testthat/test-utils.R [hunk 5]
```diff
@@ -100,3 +100,104 @@ test_that("build_app_title builts a valid tag", {
...
+    "Assertion on 'modules' failed: Must inherit from class 'teal_modules'"
+  )
+})
+
+testthat::test_that("create_app_id returns a character string", {
+  testthat::expect_true(
+    checkmate::test_string(
+      create_app_id(teal.data::teal_data(), modules(example_module()))
+    )
+  )
```
Comment (averissimo) [inline tests/testthat/test-utils.R:135]: All of `checkmate::{assert/check/test}_*` assertions have an equivalent `expect_*`

```suggestion
  checkmate::expect_string(
    create_app_id(teal.data::teal_data(), modules(example_module()))
  )
```

**Review by averissimo — APPROVED** [review]
:heart: the tests!

LGTM! 

I left a minor comment (duplicated). Feel free to resolve as you see best

File: tests/testthat/test-utils.R (hunk 5 above)
Comment (chlebowa) [reply tests/testthat/test-utils.R:135]: Excellent!

**Pull request merged**